lib/integrations/samlidp/gcpworkforce.go

```diff
@@ -0,0 +1,215 @@
+ */
+
+package samlidp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/api/googleapi"
+	iam "google.golang.org/api/iam/v1"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/integrations/samlidp/samlidpconfig"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// GCPWorkforceService defines workforce service configuration
+// parameters.
+type GCPWorkforceService struct {
+	// APIParams holds basic input params required to create GCP workforce
+	// pool and pool provider.
+	APIParams samlidpconfig.GCPWorkforceAPIParams
+	// HTTPClient is used to fetch metadata from the SAMLIdPMetadataURL
+	// endpoint.
+	HTTPClient *http.Client
+}
+
+// NewGCPWorkforceService creates a new GCPWorkforceService with input validation.
+func NewGCPWorkforceService(cfg GCPWorkforceService) (*GCPWorkforceService, error) {
+	newGCPWorkforceService := &GCPWorkforceService{
+		APIParams: samlidpconfig.GCPWorkforceAPIParams{
+			PoolName:           cfg.APIParams.PoolName,
+			PoolProviderName:   cfg.APIParams.PoolProviderName,
+			OrganizationID:     cfg.APIParams.OrganizationID,
+			SAMLIdPMetadataURL: cfg.APIParams.SAMLIdPMetadataURL,
+		},
+		HTTPClient: cfg.HTTPClient,
+	}
+
+	if err := newGCPWorkforceService.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return newGCPWorkforceService, nil
+}
+
+// CreateWorkforcePoolAndProvider creates GCP Workforce Identity Federation pool and pool provider
+// with the given GCPWorkforceAPIParams values.
+func (s *GCPWorkforceService) CreateWorkforcePoolAndProvider(ctx context.Context) error {
+	slog.InfoContext(ctx, "Configuring Workforce Identity Federation pool and SAML provider.")
+
+	iamService, err := iam.NewService(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	workforceService := iam.NewLocationsWorkforcePoolsService(iamService)
+
+	slog.With("pool_name", s.APIParams.PoolName).InfoContext(ctx, "Creating workforce pool.")
+	poolFullName := fmt.Sprintf("locations/global/workforcePools/%s", s.APIParams.PoolName)
+	createPool := workforceService.Create(
+		"locations/global",
+		&iam.WorkforcePool{
+			Name:        poolFullName,
+			DisplayName: s.APIParams.PoolName,
+			Description: "Workforce pool created by Teleport",
+			Parent:      fmt.Sprintf("organizations/%s", s.APIParams.OrganizationID),
+		})
+	createPool.WorkforcePoolId(s.APIParams.PoolName)
+	resp, err := createPool.Do()
+	if err != nil {
+		// TODO(sshah): parse through error type for better error handling
+		return trace.Wrap(err)
+	}
+
+	slog.InfoContext(ctx, "Pool created.")
+	if !resp.Done {
+		// 2 minutes timeout is semi-random decision, chosen based on the fact that
+		// when creating workforce pool from the GCP web console, it mentions
+		// that the operation could take up to 2 minutes.
+		pollCtx, cancel := context.WithTimeout(ctx, 2*time.Minute)
```

Comment:
I think 2 minutes is too short. I'd bump it to 5 minutes maybe.

Also, what happens if it fails to create within this time? What would a user do? Can they re-run the script?

Reply:
What if the pool has actually been created but it took longer than 5 minutes? Will the command pick it up and proceed to create the provider?

Reply:
I think the ideal solution here would be to have a `--pool` or `--pool-provider` flag so user can then manually create one or the other rather than having to delete and recreate. I will revisit this UX in a followup.

Reply:
(Oh Github is not showing me new comments and then suddenly I have two new comments!)

> What if the pool has actually been created but it took longer than 5 minutes? Will the command pick it up and proceed to create the provider?

Right now - the command wont pick it up once its timed out, it exits. 

> Also, in your original bash script PR, we would just execute the "create pool" command and then proceed to creating provider right away. Does "gcloud" wait for the pool to be created before returning?

Yup, `gcloud` command exits only after pool is created and is available.

Reply:
I have added 409 checker (Google api uses 409 response for conflicting resource name) if pool already exists so user have a slightly better ux in case the command timed out after creating pool, they can re run the command again and continue with pool provider. https://github.com/gravitational/teleport/pull/39820/commits/e4fad3c2858babca280a08fe103be6efc4af23c2